### DISCUSSION TITLE
What is the difference of using initializeApollo on getStaticProps

### DISCUSSION THREAD

**Discussion by theprobugmaker** [vercel/next.js#14104] [Help] [answered]
I saw this on the `api-routes-apollo-server-and-client` example

https://github.com/vercel/next.js/blob/89fffa694dc238eb40bd52809c8cbcbc39c2ec8b/examples/api-routes-apollo-server-and-client/pages/index.js#L32-L44

But on the `api-routes-apollo-server-and-client-auth` example it doesn't exist.

But in both in the `_app.js`:

https://github.com/vercel/next.js/blob/89fffa694dc238eb40bd52809c8cbcbc39c2ec8b/examples/api-routes-apollo-server-and-client/pages/_app.js#L4-L12

What about the usage of `getServerSideProps` for example to redirect the user instead of doing like this in the auth example:

https://github.com/vercel/next.js/blob/89fffa694dc238eb40bd52809c8cbcbc39c2ec8b/examples/api-routes-apollo-server-and-client-auth/pages/index.js#L22-L49

What are the differences of doing the redirect directly in the page, calling `initialApolloState` on the `getStaticProps` and using `getServerSideProps` in this situation?

**Comment by theprobugmaker** [comment]
@lfades Sorry for pinging you here but I guess that you were the one that made the example. If you can explain I would appreciate a lot, it would help me a lot.

**Comment by lfades** [comment, marked as answer]
Hi @zefexdeveloper this is a very good question. I'll try to give you a very detailed answer 😌 

`api-routes-apollo-server-and-client` is using `getStaticProps` in `pages/index.js` to add [Static Generation](https://nextjs.org/docs/basic-features/data-fetching#getstaticprops-static-generation) to the page, this is good because the data for the page in this case is known at build time, therefore we can pregenerate the page and serve the same static files to the users using the site.

> The page in question seems to be rendering user info, but that example doesn't have authentication, we could probably just update it for a more clear use case.

About `_app.js`, in both examples, and in all Apollo examples, takes care of starting the global Apollo client that can be used for all pages, it will also merge initial state that may be returned by a page using a data fetching method, like `getStaticProps` / `getServerSideProps`, but without forcing you do use them, so it will basically work with any kind of page. That's why you can see that one example is using `getStaticProps` while the other one is not, you're free to handle data in whatever way you want.

Now why is `api-routes-apollo-server-and-client-auth` not using `getStaticProps`?
Because the data is not known at build time, it depends on user info that we don't know and can be very different depending on who is using the site, so the best thing to do here is to wait for the page to be rendered, and then using client side data fetching (in this case with [SWR](https://swr.now.sh/)) we add the user data. Because the initial HTML doesn't provide this data, you can still serve static files that will be the same for anyone using the site, and are very fast, and then handle a loading state while the user data is being fetched

**Reply by theprobugmaker** [reply]
@lfades Amazing answer, thank you very much (muchas gracias) :heart:

One last question tho: having the user redirect on the page like you did on https://github.com/vercel/next.js/blob/89fffa694dc238eb40bd52809c8cbcbc39c2ec8b/examples/api-routes-apollo-server-and-client-auth/pages/index.js#L22-L49 will render the page and then redirect the user right? In my case if I want to redirect the user before he even touches that page I would need to go with `getServerSideProps` and redirect him there, am I right?

**Reply by lfades** [reply]
I was going to answer that one too in the same comment but I pressed enter accidentally 😓 

It's okay to use `getServerSideProps` and get the initial HTML with the user info, we try not to do this in our examples for the following reasons:

It's slower to wait for a serverless function to execute and give you the initial HTML, you'll have to wait more staring at a blank page, it's better to give a fast response, and then show a loading state, the user won't think the page is broken 👍 

It's also common to have data that is known at build time, and also user info, consider a blog that has a navbar with your user info, you could use SSG to generate the blog, and then after the initial HTML is received, use client side data fetching to query for the user and show the nav some moments later, we do this for https://vercel.com/. There's no need to delay the entire page just for missing data that's only required in a small portion of the page.

On another note, if your API ever goes doing, or a third party API that you use to get data breaks, your site will continue to work because the pages were already generated 🥂

**Reply by theprobugmaker** [reply]
To be honest its all new to me working with SSR and specially the SSG, I got the idea but it's still kinda confusing sometimes specially when you are not used to things that way. For example, I will explain what I'm trying to do.

I'm using using Next.js without the API routes, I know it's really good and came to help with a lot of things but it doesn't solve everything, and even before that I started creating a server with Node using TypeScript, TypeORM, TypeGraphQL, BullMQ (for queues) and things like that, authentication is working and I was able to make it work with Next.js, it was pretty easy as the server was just sending a cookie and it was being saved on the browser and I could send it using `fetch`, just like the `with-passport` example but without the server portion from the example.

The thing is, the way I used to do before is having like a middleware where when the user tries to enter in a page like `/account` which is the page that he can change things he would be redirect even before he enters the page cause he is not authenticated. In Next I saw that it doesn't happen like that (at least on the examples), most of the redirects happen on the client side. The `with-passport` for example has a "problem" that when you enter on the `profile` page it will show the heading `Profile` but will redirect you right after, it feels like a "bug". In your example at least you made a condition using `if` to check before rending the actual page.

In my case where I have a page that is protected, that needs the user to be authenticated, do you recommend using the client side rendering as you are doing or using `getServerSideProps`?

In terms of repetition, what about multiple pages, instead of writing the same conditions over and over again what would be the best way to refactor it?

Thank you very much (and Vercel, give that man a raise :laughing: )

**Reply by lfades** [reply]
Consider something like Twitter, when you refresh on Twitter you can see for a moment that the entire page is a loading state, that's client side rendering, then once the use data is loaded, you are free to render more content or redirect the user, that's usually what we recommend.

"protected" pages will work the same way whether you use client side rendering or SSR with `getServerSideProps`, there's no way to protect the code for the page with any method, the JS and or HTML that you generate for a page is public, users will always have access to that, what you can do is not give away any data that should only be accessible to a particular user, but that's never gonna happen because in the case of SSR, you'll never return the HTML and instead redirect the user, and in the case of client side rendering you'll never receive the data from the API if the user is not authenticated, and you can redirect right after.

**Reply by lfades** [reply]
About the last question on repetition, that depends on your needs, I would create an util react hook that would do that for me and put it on the pages that require user info

**Reply by theprobugmaker** [reply]
Thank you very much for the explanation. helped a lot, seriously. About the hook I think I saw an `useUser` on the `with-passport` example so I assume that is what you mean.

**Reply by lfades** [reply]
Yes 👍

**Reply by Albert-Gao** [reply]
I think I am picking up the new approach here:

Previously with getInitialProps, the lifecycle (for a full refresh) is 
1. client sends a request
1. server SSR with data
1. client renders the response (layout included because of SSR)
1. wait for JS to be initialized in order to interact with the page

So now the suggested way is:
1. client sends a request
1. server sends back the layout without data
1. client renders the response (layout included because of SSR)
1. wait for JS to be initialized in order to interact with the page
1. client sends another request in order to get data to render the page

TBH, this method is nearly as same as the setup of a PWA without SSR (like Twitter is using react native web and PWA without SSR as I am aware of), I can still see 2 benefits here:
- render after refreshing is arguably faster but depends on network speed
- no TTI problem which can not be solved on SSR-with-data, but SPA has no TTI problem as well

It should visually faster than the traditional SSR-with-data, because it is faster

then the question is what is the suggested way of returning the layout (let's say return a page with loading spinner) only without data? How could getStaticProps solve the problem?

@lfades sorry for the chime in, just want to understand more about the getStaticProps as I still have questions unanswered from that removing-getInitialProps-example-issue :D

**Reply by lfades** [reply]
> then the question is what is the suggested way of returning the layout (let's say return a page with loading spinner) only without data? How could getStaticProps solve the problem?

@Albert-Gao Not sure I understand your question, but in any case, `getStaticProps` will not have a loading state unless `getStaticPaths` is added and it returns `fallback` as `true`, in any other case `getStaticProps` will always generate the content at build time.

Pages without `getStaticProps` will also be static, you can handle the loading state in any way you want.

**Reply by sushilbansal** [reply]
> `api-routes-apollo-server-and-client` is using `getStaticProps` in `pages/index.js` to add [Static Generation](https://nextjs.org/docs/basic-features/data-fetching#getstaticprops-static-generation) to the page, this is good because the data for the page in this case is known at build time, therefore we can pregenerate the page and serve the same static files to the users using the site.

It is a user info and depends upon the whether user is logged in or not? How we can statically generate any page which is dependent on user info?

**Reply by lfades** [reply]
@sushilbansal You can opt for generating static HTML for the data of the page that's not tied to the user (like a blog post), and then fetch user related data on the client using something like SWR (https://swr.vercel.app/), like your user profile in the nav bar.
